### ISSUE TITLE
Secret and ConfigMap should limit # of keys

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#19969]
I think it's best to say "max 1000 keys, max key size is 256 characters. max total size of keys + data is 1 MB" or similar.

@pmorie promises to do this unless someone gets there first.

**Comment by bgrant0607** [comment]
Why not just aggregate size limit?

**Comment by bgrant0607** [comment]
I want a simple, configurable limit of what matters, which is object size.

**Comment by thockin** [comment]
Aggregate size limit is OK too, though I think it would be less obvious.  I
was trying to hit something succinct and memorable and not practically
limiting.

**Comment by thockin** [comment]
But this is not really a proxy for object size - all the other fields need
to be considered, and that is WAY less obvious.

**Comment by smarterclayton** [comment]
Is this fixed?

**Comment by sjenning** [comment]
No, the PR I opened didn't address the problem well so I closed it.  It really should happen at a higher level since the size limit applies to all resources stored in etcd and the size limit needs to have to serialized version of the resource to check the size.

**Comment by smarterclayton** [comment]
Why would we not perform the same validation / rules we perform for annotations, but applied to both keys and annotations?

**Comment by pmorie** [comment]
@smarterclayton If I remember correctly, we decided the approach that was needed was to do a validation on the serialized size of the ConfigMap, but there was not a good way to do it / we decided it wasn't a must-have for 1.2 / other things trumped solving it.

**Comment by smarterclayton** [comment]
Serialized size is going to be within margin of error of labels +
annotations + data.

**Comment by spiffxp** [comment]
/sig apps
I'm under the impression sig-apps owns configmaps

**Comment by kow3ns** [comment]
I think api machinery is probably a better home for this unless @bgrant0607 or @smarterclayton or @michelleN feel otherwise.

**Comment by smarterclayton** [comment]
Yes.

**Comment by gnufied** [comment]
btw - why do we only measure value lengths when measuring total size? I know serialized size still could be different but at very minimum we should be summing up both key and values whenever applicable? cc @pmorie

**Comment by liggitt** [comment]
If these APIs were new, we'd definitely want to bound key counts. At this point, I'm less sure about tightening them.

**Comment by thockin** [comment]
Agree.  We could issue warnings, but is it really a warning if it works and isn't planned to break.
